L1Trigger/TrackerTFP/src/KalmanFilter.cc

```diff
@@ -0,0 +1,398 @@
+        for (int layer : state->hitPattern().ids(false))
+          streamsStubs[offset + layer].emplace_back(TTDTC::Frame());
+      }
+    };
+    for (int channel = 0; channel < dataFormats_->numChannel(Process::kf); channel++) {
+      deque<State*> stream;
+      deque<State*> lost;
+      // proto state creation
+      int trackId(0);
+      for (TrackKFin* track : input_[channel]) {
+        State* state = nullptr;
+        if (track) {
+          states_.emplace_back(dataFormats_, track, trackId++);
+          state = &states_.back();
+        }
+        stream.push_back(state);
+      }
+      // state propagation
+      for (layer_ = 0; layer_ < setup_->numLayers(); layer_++)
+        layer(stream);
+      // untruncated best state selection
+      deque<State*> untruncatedStream = stream;
+      accumulator(untruncatedStream);
+      // apply truncation
+      if (enableTruncation_ && (int)stream.size() > setup_->numFrames())
+      //if (enableTruncation_ && (int)stream.size() > setup_->numFramesIO())
+        stream.resize(setup_->numFrames());
+      // best state per candidate selection
+      accumulator(stream);
+      deque<State*> truncatedStream = stream;
+      // storing of best states missed due to truncation
+      sort(untruncatedStream.begin(), untruncatedStream.end());
+      sort(truncatedStream.begin(), truncatedStream.end());
+      set_difference(untruncatedStream.begin(), untruncatedStream.end(), truncatedStream.begin(), truncatedStream.end(), back_inserter(lost));
+      // store found tracks
+      put(stream, acceptedStubs, acceptedTracks, channel);
+      // store lost tracks
+      put(lost, lostStubs, lostTracks, channel);
+    }
+  }
+
+  // adds a layer to states
+  void KalmanFilter::layer(deque<State*>& stream) {
+    static constexpr int latency = 5;
+    // dynamic state container for clock accurate emulation
+    deque<State*> streamOutput;
+    deque<State*> stack;
+    // static delay container
+    vector<State*> delay(latency, nullptr);
+    delay.reserve(latency + 1);
+    // each trip corresponds to a f/w clock tick
+    // done if no states to process left, taking as much time as needed
+    while (!stream.empty() || !stack.empty() || !all_of(delay.begin(), delay.end(), [](const State* state){ return state == nullptr; })) {
+      State* state = pop_front(stream);
+      if (!state)
+        state = pop_front(stack);
+      streamOutput.push_back(state);
+      if (!state || !state->stub() || state->layer() != layer_)
+        state = nullptr;
+      if (state != nullptr)
+        comb(state);
+      delay.push_back(state);
+      state = pop_front(delay);
+      if (state != nullptr)
+        stack.push_back(state);
+    }
+    stream = streamOutput;
+    for (State*& state : stream) {
+      if (!state || !state->stub() || state->layer() != layer_ )
+        continue;
+      update(state);
+    }
+  }
+
+  // repicks combinatoric stubs for state
```

Comment:
Change comment "Find next combinatoric (i.e. not first in layer) stub for state"

Reply:
done